Here is how you can load the datasets with S3
1. In Config file: LOAD_FROM_S3=True
2. Enter you SSP Cloud secrets in a .env file (see READme)

In [1]:
#Imports
import s3fs
import pandas as pd
import logging
logging.basicConfig(level=logging.INFO)  #to be able to see the logging info+warnings(to only see warnings: logging.DEBUG)

from src.data_processing.data_load import data_loading
from src.data_processing.data_processing import DataProcessor

In [3]:
#To load data the from the different files directly
df_portefeuille, df_consommations, df_reclamations, df_interactions, df_impayes= data_loading("training")

/home/onyxia/work/Alptis-churn-prediction/src/data_processing/data_load.py:19: DtypeWarning: Columns (31,33) have mixed types. Specify dtype option on import or set low_memory=False.
  data_frames[key] = pd.read_csv(filepath, sep=";", index_col=False)


In [2]:
#To load and process the data of a specific set (training, validation, evaluation) without normalisation of numerical values
processor = DataProcessor(mode="training")
df = processor.run(optional_normalisation=False)

INFO:src.data_processing.data_processing:Dropped columns from dataset: ['client_structure_familiale', 'client_nom_banque', 'client_nps_verbatim_n', 'client_nps_verbatim_n_moins1', 'impaye_nb_actions', 'impaye_montant_total', 'impaye_duree_max_action_jours', 'impaye_action_1er_impaye', 'impaye_action_1ere_lettre_relance', 'impaye_action_2eme_impaye', 'impaye_action_annonce_contentieux', 'impaye_action_en_recouvrement', 'impaye_action_mise_en_demeure', 'impaye_action_suspension_garanties', 'interaction_historique_mail', 'client_nps_date_reponse_n_moins1_jours', 'client_nps_date_reponse_n_jours']
INFO:src.data_processing.data_processing:Columns that are One-hot encoded: ['client_mode_paiement', 'courtier_reseau_courtage', 'courtier_segmentation_interne', 'courtier_type_commission', 'recla_canal_entrant_principale', 'recla_canal_entrant_secondaire']
INFO:src.data_processing.data_processing:Columns that are Target encoded: ['client_ro_souscripteur', 'client_produit', 'client_garantie_base',

In [7]:
processor = DataProcessor(mode="training")
processor.manual_preprocessing(df_portefeuille, df_consommations, df_reclamations, df_interactions, df_impayes)

INFO:src.data_processing.data_processing:Dropped columns from dataset: ['client_structure_familiale', 'client_nom_banque', 'client_nps_verbatim_n', 'client_nps_verbatim_n_moins1', 'impaye_nb_actions', 'impaye_montant_total', 'impaye_duree_max_action_jours', 'impaye_action_1er_impaye', 'impaye_action_1ere_lettre_relance', 'impaye_action_2eme_impaye', 'impaye_action_annonce_contentieux', 'impaye_action_en_recouvrement', 'impaye_action_mise_en_demeure', 'impaye_action_suspension_garanties', 'interaction_historique_mail']


,client_code,client_age_souscripteur,client_ro_souscripteur,client_code_postal,client_departement,client_a_conjoint,client_nb_enfants,client_nb_assures_contrat,client_a_garantie_prevoyance,client_produit,...,interaction_nb_transferts_reclamation,interaction_nb_transferts_gestion,interaction_nb_transferts_commercial,interaction_nb_transferts_externalises,age_categories,courtier_anciennete_categories,client_nps_date_reponse_n_moins1_jours,client_nps_date_reponse_n_jours,client_date_debut_effet_garantie_mois,dernier_paiement_consommation_mois
0,21736,97,SS,69100,69,0.0,0.0,1.0,0,produit_21,...,0.0,0.0,0.0,0.0,senior plus,very old,-1.0,-1.0,766,1.0
1,22037,95,SS,26230,26,0.0,0.0,1.0,1,produit_21,...,0.0,1.0,0.0,1.0,senior plus,old,-1.0,-1.0,322,1.0
2,22331,88,TNS,69760,69,1.0,0.0,2.0,0,produit_21,...,0.0,5.0,0.0,7.0,senior plus,very old,-1.0,-1.0,740,1.0
3,22604,95,TNS,69003,69,0.0,0.0,1.0,0,produit_21,...,0.0,0.0,0.0,0.0,senior plus,old,-1.0,-1.0,359,1.0
4,22702,90,TNS,69100,69,1.0,0.0,2.0,1,produit_21,...,0.0,4.0,0.0,8.0,senior plus,very old,-1.0,-1.0,727,1.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
132571,16204966,36,SS,74140,74,0.0,0.0,1.0,0,produit_32,...,0.0,1.0,0.0,0.0,mature,old,-1.0,-1.0,1,-1.0
132572,16205008,63,SS,17443,17,0.0,0.0,1.0,0,produit_20,...,0.0,0.0,0.0,0.0,senior,stable,-1.0,-1.0,1,-1.0
132573,16205281,25,SS,74200,74,0.0,0.0,1.0,0,produit_32,...,0.0,0.0,0.0,0.0,young,new,-1.0,-1.0,1,-1.0
132574,16206464,22,SS,97215,97,0.0,0.0,1.0,1,produit_18,...,0.0,0.0,0.0,0.0,young,stable,-1.0,-1.0,1,-1.0


In [10]:
#Same but with normalisation of numerical values
processor = DataProcessor(mode="training")
df_norm = processor.run(optional_normalisation=True)

INFO:src.data_processing.data_processing:Dropped columns from dataset: ['client_structure_familiale', 'client_nom_banque', 'client_nps_verbatim_n', 'client_nps_verbatim_n_moins1', 'impaye_nb_actions', 'impaye_montant_total', 'impaye_duree_max_action_jours', 'impaye_action_1er_impaye', 'impaye_action_1ere_lettre_relance', 'impaye_action_2eme_impaye', 'impaye_action_annonce_contentieux', 'impaye_action_en_recouvrement', 'impaye_action_mise_en_demeure', 'impaye_action_suspension_garanties', 'interaction_historique_mail']
INFO:src.data_processing.data_processing:Encoded with one-hot encoding: ['client_mode_paiement', 'client_frequence_paiement', 'courtier_reseau_courtage', 'courtier_segmentation_interne', 'courtier_type_commission', 'recla_canal_entrant_principale', 'recla_canal_entrant_secondaire']
INFO:src.data_processing.data_processing:Encoded with target encoding: ['client_ro_souscripteur', 'client_produit', 'client_garantie_base', 'client_garantie_base_plus_options', 'client_millesim

In [ ]:
df_norm #now for example, age is normalised (better to be processed by a linear model OR other specific models)

0         5.0
1         5.0
2         5.0
3         5.0
4         5.0
         ... 
132571    2.0
132572    4.0
132573    0.0
132574    0.0
132575    2.0
Name: age_categories, Length: 132576, dtype: float64

**VERY IMPORTANT**

*The test/validation set need to be pre-processed using encoder/scaler that were fitted on training set* to avoid data-leakage (due to target-encoding, normalisation, ...).
To do see, see below: 

1. method df.run(): fit + transform the dataframe --> to use for training
2. method df.run_transform(): only transform the dataframe, using pre-saved args for target-encoding, scaler, ordinal encoding,... --> to use for validation/test

In [4]:
#1. Pre-process training data (scaler and encoder are saved during this process)
train_processor = DataProcessor(mode="training")
df_train = train_processor.run()  #fit training set and save arguments (scaler, ordinal map, target_encoding_maps, ...)

#2. Pre-process test data (by using the same scaler/encoder fitted on training dataset)
test_processor = DataProcessor(mode="evaluation")

#  HERE:  Reuse the same pre-processor fitted on training dataset (so no data-leakage with target-encoding, normalisation...)
test_processor.target_encoding_maps = train_processor.target_encoding_maps
test_processor.ordinal_maps          = train_processor.ordinal_maps
test_processor.encoded_columns       = train_processor.encoded_columns
test_processor.scaler                = train_processor.scaler
test_processor.normalized_columns    = train_processor.normalized_columns

df_test = test_processor.run_transform() #pre-process test data using same args 

INFO:src.data_processing.data_processing:Dropped columns from dataset: ['client_structure_familiale', 'client_nom_banque', 'client_nps_verbatim_n', 'client_nps_verbatim_n_moins1', 'impaye_nb_actions', 'impaye_montant_total', 'impaye_duree_max_action_jours', 'impaye_action_1er_impaye', 'impaye_action_1ere_lettre_relance', 'impaye_action_2eme_impaye', 'impaye_action_annonce_contentieux', 'impaye_action_en_recouvrement', 'impaye_action_mise_en_demeure', 'impaye_action_suspension_garanties', 'interaction_historique_mail', 'client_nps_date_reponse_n_moins1_jours', 'client_nps_date_reponse_n_jours']
INFO:src.data_processing.data_processing:Columns that are One-hot encoded: ['client_mode_paiement', 'courtier_reseau_courtage', 'courtier_segmentation_interne', 'courtier_type_commission', 'recla_canal_entrant_principale', 'recla_canal_entrant_secondaire']
INFO:src.data_processing.data_processing:Columns that are Target encoded: ['client_ro_souscripteur', 'client_produit', 'client_garantie_base',

In [ ]:
#Visualise saved args of training set (ex: target_encoding_maps)
train_processor.target_encoding_maps #ex: for each column target-encoded, save the target-mean of each values

{'client_ro_souscripteur': client_ro_souscripteur
 A           0.000000
 AM          0.068583
 AMEXA       0.074365
 ASSIM_SS    0.052632
 CFE         0.000000
 DDOM        0.000000
 EDOM        0.000000
 F           0.000000
 MGEN        0.125817
 MONACO      0.000000
 NOTAIRE     0.125000
 PRAGA       0.000000
 PT          0.000000
 RAM         0.042394
 SA          0.100908
 SNCF        0.140845
 SS          0.123033
 SSM         0.210526
 TNS         0.062628
 Name: target_resiliation_6mois, dtype: float64,
 'client_produit': client_produit
 produit_1     0.053705
 produit_10    0.067890
 produit_11    0.066327
 produit_12    0.112782
 produit_13    0.000000
 produit_14    0.022305
 produit_15    0.038911
 produit_16    0.052288
 produit_17    0.081159
 produit_18    0.058163
 produit_19    0.022013
 produit_2     0.071063
 produit_20    0.013863
 produit_21    0.021672
 produit_22    0.046537
 produit_23    0.059028
 produit_24    0.006623
 produit_25    0.146186
 produit_26    0.

## New external column: client_revenu_commune_2021

In [ ]:
df[["client_departement","client_code_postal","client_revenu_commune_2021"]] #new column on revenu (external)

,client_departement,client_code_postal,client_revenu_commune_2021
0,69,69100,21540.0
1,26,26230,22350.0
2,26,26230,24740.0
3,26,26230,23960.0
4,26,26230,24690.0
...,...,...,...
694135,97,97215,19840.0
694136,95,95300,29520.0
694137,95,95300,29670.0
694138,95,95300,33270.0


In [39]:
#check coherence by look at richer and poorer departement of France
(df.groupby("client_code_postal")["client_revenu_commune_2021"]
   .median()
   .nlargest(10)
   .reset_index())

,client_code_postal,client_revenu_commune_2021
0,92200,48010.0
1,01220,46650.0
2,74160,45460.0
3,01280,44430.0
4,78860,44020.0
5,01170,44010.0
6,78110,42870.0
7,69370,42270.0
8,01630,42220.0
9,92430,41740.0


In [40]:
#poorer commune
(df.groupby("client_code_postal")["client_revenu_commune_2021"]
   .median()
   .nsmallest(10)
   .reset_index())

,client_code_postal,client_revenu_commune_2021
0,97433,13280.0
1,97413,13780.0
2,97439,14030.0
3,97470,14550.0
4,59100,14580.0
5,91350,14790.0
6,09320,14905.0
7,97442,14910.0
8,97420,14970.0
9,97440,15080.0


It's coherent: 

Richest: 
- 92200 = Neuilly-sur-Seine
- 01220 = Divonne les bains (ville frontalière à la suisse)

Poorest: 
- 974... = Réunion
- 91350 = Grigny